### PR TITLE
Setting a new animal's DMI when she is added to the herd

### PR DESCRIPTION
**Pull request by militsasotirova**
When a new animal gets added to the herd on a day which the ration is not calculated, we set the ration formulation of the new animal to be the same as that of the animals in the pen she is added to. However, we don't set her dry matter intake. This PR makes that change.

### CODE REVIEW

**Comment by militsasotirova** [comment]
Another issue when we add a new animal to a pen has to do with the `ration` attribute of the Pen class. Once a ration is formulated for the average animal in the pen, the ration attribute is set to be [that ration multiplied by the number of animals in the pen](https://github.com/RuminantFarmSystems/MASM/blob/a449ddd0d4bf34518737f62b999d2e11054a6033/RUFAS/routines/animal/pen.py#L435) so it represents the total ration for the pen. When we add a new animal to the herd within a ration interval, her ration becomes the (total ration of the pen she was added to) / (number of animals in the pen before she was added). This isn't a problem when we add 1 animal in a day, but if we add more than 1, the (number of animals in the pen before she was added) changed because we have already added another animal, making the division inaccurate since the (total ration) no longer reflects the ration of ALL animals in the pen. The solution was to store the amounts of animals in each pen before ANY additions and use those numbers when assigning each new animal's ration. Then, we also have to adjust the pen's ration attribute to reflect the new number of animals in the pen (this is also necessary logic when there is only 1 animal added to the herd). The commit I just made fixes this. Similar logic applies to the `manure` attribute of the Pen class.

Since the changes in this PR were made because of discrepancies found within [a different branch](https://github.com/RuminantFarmSystems/MASM/pull/106) between the total herd DMI and the total purchased feed, it is not easy in this branch to see that these changes fix that discrepancy. If you wanted to do so here, you could insert something like the following code snippet in `simulation_engine.py` in the `daily_simulation()` function after the daily routines are called and before the output updates.

```
herd_dmi = 0
purchased_feed_total = 0
for animal in state.animal_management.heiferIs + \
              state.animal_management.heiferIIs + \
              state.animal_management.heiferIIIs + \
              state.animal_management.cows:
    herd_dmi += animal.dry_matter_intake

    for key in animal.ration_formulation:
        if key != "status" and key != "objective":
            purchased_feed_total += animal.ration_formulation[key]

assert abs(purchased_feed_total - herd_dmi) < 0.5 # 0.5 here is arbitrary - the purpose is to assert that the difference is small 
```

The reason that the `calves` list is not included in the for-loop is that since a calf's diet consists partly of non-purchased feed (i.e. the milk from her mother) and partly of purchased feed (like starter), there can never be a balance between a calf's dry matter intake and her purchased feed total since the dry matter intake is calculated from her whole diet. I have more information in the comment found in [this commit to a different branch.](https://github.com/RuminantFarmSystems/MASM/pull/106/commits/bbb01a1f24b4dcd0e60c2781cd48876fda38fecd)

**Review by KFosterReed — APPROVED** [review]
These changes still work with the current json files and fix the discrepancies between the purchased feed amounts and intake. I cross referenced the changes with the incoming changes in animal_management.py and pen.py from the heifer ration branch and they don't look like they will conflict with the updates Chris made to the pen management and animal allocation processes. Looks good to me!

**Comment by KFosterReed** [comment]
I added a few other people as suggested reviewers to see if we can get this moving :)

File: RUFAS/routines/animal/pen.py [hunk 1]
```diff
@@ -199,6 +202,7 @@ def __init__(self, id_number, vert_dist, horiz_dist, num_stalls, housing_type,
                                        'NDF': 0, 'lignin': 0, 'ash': 0,
                                        'phosphorus': 0, 'potassium': 0, 'N': 0}
         self.ration_nutrient_conc = {}
+        self.dry_matter_intake = 0
```
Comment (manfei-L) [inline RUFAS/routines/animal/pen.py]: Should it be initialized as a float with 0.0?

File: RUFAS/routines/animal/pen.py [hunk 2]
```diff
@@ -561,13 +566,14 @@ def set_up_new_animal(self, animal, p_comp, feed, temp):
                 calf and that its p_animal attribute has already been calculated
             feed: instance of the Feed class defined in feed.py
             temp: temperature on the current day
+            num_animals_before_additions: the number of animals in this pen
+                before any new animals were added for the day
         """
-        num_animals = len(self.animals_in_pen)
-        if num_animals == 0:
+        if num_animals_before_additions == 0:
```
Comment (manfei-L) [inline RUFAS/routines/animal/pen.py:572]: Curious: could an animal be added to an empty pen?

**Review by manfei-L — COMMENTED** [review]
These changes are important for representing the cases when a new animal is added to the herd. The code and descriptions look good to me. I noticed one potential issue in the init of a variable (dry_matter_intake), initiated as int while described in the documentation above as float. Other than that, I think we can merge this pull request.

File: RUFAS/routines/animal/pen.py (hunk 2 above)
Comment (militsasotirova) [reply RUFAS/routines/animal/pen.py:572]: Yes, I encountered a bug in the past where the pens started out as populated and at one point one of the pens ended up with a population of 0. There were issues (I forget the details) when an animal was added to that pen after.

**Review by manfei-L — APPROVED** [review]
The changes look good to me in this pull request.

**Pull request merged**